Modules/LearningSequence/classes/class.ilLSLocalDI.php

```diff
@@ -0,0 +1,234 @@
+<?php declare(strict_types=1);
+
+use Pimple\Container;
+
+/**
+  * @author Nils Haagen <nils.haagen@concepts-and-training.de>
+ */
+trait ilLSLocalDI
+{
+	public function getLSLocalDI(
+		ilObjLearningSequence $object,
+		ArrayAccess $dic
+	): Container {
+		$container = new Container();
+
+		$ref_id = (int)$object->getRefId();
+		$obj_id = (int)$object->getId();
+		$obj_title = $object->getTitle();
+
+		$current_user = $dic['ilUser'];
+		$current_user_id = (int)$current_user->getId();
+
+		$data_factory = new \ILIAS\Data\Factory();
+
+		$container["obj.ref_id"] = $ref_id;
+		$container["obj.obj_id"] = $obj_id;
+		$container["obj.title"] = (string)$obj_title;
+
+		$container["usr.id"] = $current_user_id;
+
+		$container["obj.sorting"] = function($c): ilContainerSorting
+		{
+			return ilContainerSorting::_getInstance($c["obj.obj_id"]);
+		};
+
+		$container["db.filesystem"] = function($c): ilLearningSequenceFilesystem
+		{
+			 return new ilLearningSequenceFilesystem();
+		};
+
+		$container["db.settings"] = function($c) use ($dic): ilLearningSequenceSettingsDB
+		{
+			return new ilLearningSequenceSettingsDB(
+				$dic["ilDB"],
+				$c["db.filesystem"]
+			);
+		};
+
+		$container["db.activation"] = function($c) use ($dic): ilLearningSequenceActivationDB
+		{
+			return new ilLearningSequenceActivationDB($dic["ilDB"]);
+		};
+
+		$container["db.states"] = function($c) use ($dic): ilLSStateDB
+		{
+			return new ilLSStateDB($dic["ilDB"]);
+		};
+
+		$container["db.postconditions"] = function($c) use ($dic): ilLSPostConditionDB
+		{
+			return new ilLSPostConditionDB($dic["ilDB"]);
+		};
+
+		$container["db.lsitems"] = function($c) use ($dic): ilLSItemsDB
+		{
+			$online_status = new LSItemOnlineStatus();
+			return new ilLSItemsDB(
+				$dic["tree"],
+				$c["obj.sorting"],
+				$c["db.postconditions"],
+				$online_status
+			);
+		};
+
+		$container["db.progress"] = function($c) use ($dic): ilLearnerProgressDB
+		{
+			 return new ilLearnerProgressDB(
+				$c["db.lsitems"],
+				$dic["ilAccess"]
+			);
+		};
+
+		$container["learneritems"] = function($c): ilLSLearnerItemsQueries
+		{
+			return new ilLSLearnerItemsQueries(
+				$c["db.progress"],
+				$c["db.states"],
+				$c["obj.ref_id"],
+				$c["usr.id"]
+			);
+		};
+
+
+		$container["gui.learner"] = function($c) use ($dic, $object): ilObjLearningSequenceLearnerGUI
+		{
+			$has_items = count($c["learneritems"]->getItems()) > 0;
+			$first_access = $c["learneritems"]->getFirstAccess();
+
+			return new ilObjLearningSequenceLearnerGUI(
+				$c["obj.ref_id"],
```

Comment:
Das ist problematisch, denn nun kann man diesen dic nur für diese ref_id Benutzen oder muss immer wieder einen neuen dic generieren, welcher dann auch alle dependencies neu erzeugen muss, was sehr unperformat ist. Ferner hat das dic einen internen Zustand, den man umständlich auslesen muss. Ich sehe bei Pimple keine möglichkeit einen Parameter beim Container-Aufruf zu übergebn. Kann man die Ref-Id nach der Erzeugung definieren?

Reply:
Beziehst Du Dich hier auf diese eine, konkrete Zeile oder den gesamten DIC?

Reply:
Ich beziehe mich hier allgemein darauf, dass du eine ref-id als DIC parameter nutzt und diese dann an all Konsumente darunter vererbt wird, so wie speziell in Modules/LearningSequence/classes/class.ilLSLocalDI.php: 100.